### PR TITLE
Make base toloka object children genericable

### PR DESCRIPTION
**Pull request by Pocoder**
Make classes that are inherited from Generic and BaseTolokaObject structureable. 

## Description
I propose to add one more structure hook that is called when type is generic and it has structure method.
I use functions from `cattr.gen` to get mapping from TypeVar to resolved types. New func `get_mapped_type` is used to resolve whole types. Then these types substitute for class attrs and `structure` method of class after substitution is called.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation and examples improvement (changes affected documentation and/or examples)

## Checklist:

- [x] I have read the **CONTRIBUTING** document.
- [x] I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: src/client/primitives/base.py [hunk 1]
```diff
@@ -314,10 +331,14 @@ def structure(cls, data: Any):
 
             value = data.pop(key)
             if field.type is not None:
-                value = converter.structure(value, field.type)
+                # if isinstance(field.type, typing.TypeVar) and field.type.__name__ in type_var_mapping:
+                #     target_type = type_var_mapping[field.type.__name__]
+                # else:
+                #     target_type = field.type
```
Comment (alexdrydew) [inline src/client/primitives/base.py:337]: There is commented code, should we remove it?

File: tests/primitives/test_base.py [hunk 2]
```diff
@@ -19,6 +21,26 @@ def test_base_toloka_object_is_pickle_serializable(base_toloka_object):
     assert deserialized == base_toloka_object
 
 
+def test_generic_base_inherited_object():
```
Comment (alexdrydew) [inline tests/primitives/test_base.py:24]: Can you please add more tests? In particular:
* Generic with 2+ typevars
* Generic with nested generic
* Test with nongeneric subclass of generic:
```
class A(Generic[T]):
    a: T
class B(A[int]):
    b: str
```

File: src/client/_converter.py [hunk 3]
```diff
@@ -11,21 +11,21 @@
...
+
 if _CATTRS_VERSION < (22, 2, 0):
     converter = cattr.Converter()
 else:
     converter = cattr.converters.BaseConverter()
 
 converter.register_structure_hook_func(
     lambda type_: hasattr(type_, 'structure'),
-    lambda data, type_: type_.structure(data)  # type: ignore
+    lambda data, type_: getattr(type_, 'structure').__func__(type_, data)  # type: ignore
```
Comment (alexdrydew) [inline src/client/_converter.py:22]: Why this is required?

File: src/client/primitives/base.py [hunk 4]
```diff
@@ -285,6 +292,16 @@ def unstructure(self) -> Optional[dict]:
...
 
+        type_var_mapping = {}
+        if is_generic(cls):
+            base = get_origin(cls)
+            type_var_mapping = _generate_mapping(cls, type_var_mapping)
+            cls = base
+        for base in getattr(cls, "__orig_bases__", ()):
+            if is_generic(base) and not str(base).startswith("typing.Generic"):
+                type_var_mapping = _generate_mapping(base, type_var_mapping)
+                break
```
Comment (alexdrydew) [inline src/client/primitives/base.py:303]: Let's reimplement `_generate_mapping` in toloka.util, it is risky to depend on private API in core logic. Probably all this code block can be extracted to a function

File: src/client/primitives/base.py [hunk 5]
```diff
@@ -337,6 +358,36 @@ def from_json(cls, json_str: str):
...
+    try:
+        origin_type = t.__dict__['__origin__']
+        type_args = t.__dict__.get('__args__', [])
+    except AttributeError:
+        return t
+    except KeyError:
+        return t
+    mapped_args = tuple(_get_mapped_type(t_arg, mapping) for t_arg in type_args)
+
+    # TODO: support all generic types including custom generics
```
Comment (alexdrydew) [inline src/client/primitives/base.py:373]: Are nested attrs generic supported?

File: src/client/_converter.py (hunk 3 above)
Comment (Pocoder) [reply src/client/_converter.py:22]: I was not looking into details, but I think classmethod decorator breaks smth, idk.
Example:
```python
from typing import Generic, TypeVar


T = TypeVar('T')


class A(Generic[T]):
    data: T

    @classmethod
    def structure(cls, data):
        print(cls)
        print(data)


if __name__ == '__main__':
    data = {'data': 1}
    A[int].structure(data)  # printed: <class '__main__.A'>
    getattr(A, 'structure').__func__(A[int], data)  # printed __main__.A[int]
```

File: src/client/primitives/base.py (hunk 5 above)
Comment (Pocoder) [reply src/client/primitives/base.py:373]: Actually yes

File: src/async_client/client.py [hunk 6]
```diff
@@ -84,7 +84,6 @@ async def _do_request_with_retries(self, method, path, **kwargs):
         @self.retrying.wraps
         async def wrapped(method, path, **kwargs):
             response = await self._session.request(method, path, **kwargs)
-            await response.aread()
```
Comment (alexdrydew) [inline src/async_client/client.py:87]: I was surprised that it works so I added a new test case: https://github.com/Toloka/toloka-kit/pull/185
Seems like it is true that this line is unnecessary

**Review by alexdrydew — APPROVED** [review]

**Pull request merged**